payu/branch.py

```diff
@@ -0,0 +1,203 @@
+        config_path = Path(DEFAULT_CONFIG_FNAME)
+        config_path.resolve()
+
+    # Check for valid paths
+    skip_msg = f"Skipping adding 'restart: {restart_path}' to config file"
+    if not config_path.exists() or not config_path.is_file:
+        warnings.warn(f"Given configuration file {config_path} does not "
+                      "exist. " + skip_msg)
+        return
+    if not restart_path.exists() or not restart_path.is_dir():
+        warnings.warn((f"Given restart directory {restart_path} does not "
+                       "exist. " + skip_msg))
+        return
+
+    # Default ruamel yaml preserves comments and multiline strings
+    yaml = YAML()
+    config = yaml.load(config_path)
+
+    # Add in restart path
+    config['restart'] = str(restart_path)
+
+    # Write modified lines back to config
+    yaml.dump(config, config_path)
+    print(f"Added 'restart: {restart_path}' to configuration file:",
+          config_path.name)
+
+
+def get_control_path(config_path):
+    """Given the config path, return the control path"""
+    # Note: Control path is set in read_config
+    config = read_config(config_path)
+    return Path(config.get('control_path'))
+
+
+def checkout_branch(lab: Laboratory,
+                    branch_name: str,
+                    is_new_branch: bool = False,
+                    is_new_experiment: bool = False,
+                    start_point: Optional[str] = None,
+                    restart_path: Optional[Path] = None,
+                    config_path: Optional[Path] = None) -> None:
+    """Checkout branch"""
+    control_path = get_control_path(config_path)
+
+    # Checkout branch
+    git_checkout_branch(control_path, branch_name, is_new_branch, start_point)
+
+    metadata = Metadata(lab, branch=branch_name, config_path=config_path)
+    if is_new_branch or is_new_experiment:
+        # Creates new uuid, experiment name, updates and commit metadata file
+        metadata.setup_new_experiment()
+    else:
+        # Create/update metadata if no uuid, otherwise run checks on existing
+        # metadata and commit any changes
+        metadata.setup()
+        metadata.commit_file()
+
+    # Add restart option to config
+    if restart_path:
+        add_restart_to_config(restart_path, config_path=config_path)
+
+    # Switch/Remove/Add archive and work symlinks
+    experiment = metadata.experiment_name
+    switch_symlink(Path(lab.archive_path), control_path, experiment, 'archive')
+    switch_symlink(Path(lab.work_path), control_path, experiment, 'work')
+
+
+def switch_symlink(lab_dir_path: Path, control_path: Path,
+                   experiment_name: str, sym_dir: str) -> None:
+    """Helper function for removing and switching work and archive
+    symlinks in control directory"""
+    dir_path = lab_dir_path / experiment_name
+    sym_path = control_path / sym_dir
+
+    # Remove symlink if it already exists
+    if sym_path.exists() and sym_path.is_symlink:
```

Comment:
Otherwise if there is no archive (e.g. its a new branch), there would be a symlink to an archive thats from a different branch. In this case I think it's better to have no archive symlink to avoid any confusion.